File: shams_cash_pulse/cash_pulse/report/purchase_forecast/purchase_forecast.js
// Copyright (c) 2026, SHAMS SOLUTIONS and contributors
/* eslint-disable */

frappe.query_reports["Purchase Forecast"] = {
    "filters": [
        {
            "fieldname": "only_to_order",
            "label": "الأصناف التي تحتاج طلباً فقط",
            "fieldtype": "Check",
            "default": 0
        },
        {
            "fieldname": "only_stockout",
            "label": "الأصناف النافدة فقط",
            "fieldtype": "Check",
            "default": 0
        }
    ],
    "formatter": function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (!data) {
            return value;
        }

        if (column.fieldname === "status") {
            if (data.status === "نافد") {
                value = `<span style="color:#fff; background:#C0392B; padding:2px 8px; border-radius:4px; font-weight:600; font-size:11px;">${value}</span>`;
            } else if (data.status === "يُطلب") {
                value = `<span style="color:#fff; background:#1E7145; padding:2px 8px; border-radius:4px; font-weight:600; font-size:11px;">${value}</span>`;
            } else {
                value = `<span style="color:#7f8c8d; font-size:11px;">${value}</span>`;
            }
        }
        if (column.fieldname === "to_order" && data.to_order > 0) {
            value = `<b style="color:#1E7145;">${value}</b>`;
        }
        if (column.fieldname === "expired_stock" && data.expired_stock > 0) {
            value = `<span style="color:#E67E22; font-weight:600;">${value}</span>`;
        }
        return value;
    }
};

File: shams_cash_pulse/cash_pulse/report/purchase_forecast/purchase_forecast.py
# Copyright (c) 2026, SHAMS SOLUTIONS and contributors
# تقرير التنبؤ الشرائي — Script Report

import frappe
from shams_cash_pulse.cash_pulse.forecast_calculator import compute_forecast


def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    raw = compute_forecast()

    # فلترة العرض (لا تمسّ الحساب)
    only_order = filters.get("only_to_order")
    only_stockout = filters.get("only_stockout")

    data = []
    for r in raw:
        if only_order and r["to_order"] <= 0:
            continue
        if only_stockout and not r["is_stockout"]:
            continue

        status = "نافد" if r["is_stockout"] else ("يُطلب" if r["to_order"] > 0 else "مكتفٍ")
        data.append({
            "item_code": r["item_code"],
            "item_name": r["item_name"],
            "status": status,
            "sold_period": int(r["sold_period"]),
            "daily_rate": float(r["daily_rate"]),
            "expected_demand": int(r["expected_demand"]),
            "bin_qty": int(r["bin_qty"]),
            "valid_stock": int(r["valid_stock"]),
            "expired_stock": int(r["expired_stock"]),
            "to_order": int(r["to_order"]),
            "rate": float(r["rate"]),
            "order_value": float(r["order_value"]),
            "explanation": r["explanation"],
        })

    return columns, data


def get_columns():
    return [
        {"label": "الكمية المطلوبة", "fieldname": "to_order", "fieldtype": "Int", "width": 120},
        {"label": "الحالة", "fieldname": "status", "fieldtype": "Data", "width": 80},
        {"label": "الصنف", "fieldname": "item_name", "fieldtype": "Data", "width": 280},
        {"label": "رقم الصنف", "fieldname": "item_code", "fieldtype": "Data", "width": 130},
        {"label": "بيع الفترة", "fieldname": "sold_period", "fieldtype": "Int", "width": 90},
        {"label": "معدّل/يوم", "fieldname": "daily_rate", "fieldtype": "Float", "width": 90, "precision": 2},
        {"label": "الطلب المتوقع", "fieldname": "expected_demand", "fieldtype": "Int", "width": 110},
        {"label": "مخزون", "fieldname": "bin_qty", "fieldtype": "Int", "width": 90},
        {"label": "الصالح", "fieldname": "valid_stock", "fieldtype": "Int", "width": 90},
        {"label": "المهدّد", "fieldname": "expired_stock", "fieldtype": "Int", "width": 90},
        {"label": "التكلفة", "fieldname": "rate", "fieldtype": "Currency", "width": 90},
        {"label": "قيمة الطلب", "fieldname": "order_value", "fieldtype": "Currency", "width": 120},
        {"label": "التوضيح", "fieldname": "explanation", "fieldtype": "Data", "width": 400},
    ]
